docs/source/developer/custom_pipeline.rst

```diff
@@ -0,0 +1,33 @@
+.. _arch-pipeline:
+
+========================
+Notes on Custom Pipeline
+========================
+
+.. warning:: The custom pipeline feature is regard as expert only.  Modifying
+             the compiler behavior can invalidate internal assumptions in the
+             numba source code.
+
+
+For library developers looking for a way to extend or modify the compiler
+behavior, you can do so by defining a custom compiler pipeline by inheriting
+from ``numba.compiler.BasePipeline``.  The default numba pipeline is defined
+as ``numba.compiler.Pipeline``.  It implements the ``.define_pipelines()``
+method, which adds the *nopython-mode*, *object-mode* and *interpreted-mode*
+pipelines.  These three pipelines are defined by ``.define_nopython_pipeline``,
```

Comment:
Perhaps `These three pipelines are defined in ``BasePipeline`` by the methods ``.define_nopython_pipeline``, ``.define_objectmode_pipeline`` and ``.define_interpreted_pipeline``,
respectively.` ?